# Signal Acquisition Planning

This notebook demonstrates the planning of signal capture and spectral analysis parameters. 

The configuration includes:
- Desired Resolution Bandwidth (RBW): 1 kHz
- Sampling Rate (fs): 20 MHz
- Bandwidth: 20 MHz
- Window Type: Hamming
- Overlap: 50%
- K (averaging factor): 1
- Size planning method: next_5smooth

The notebook uses `plan_capture_parameters()` and `plan_welch_parameters()` functions to calculate optimal FFT size and other parameters based on these requirements.


In [1]:
from rfscope.common.utils import plan_capture_parameters
from rfscope.dsp.spectral import plan_welch_parameters

In [2]:
plan = {
    'rbw_hz': 1e3,
    'fs_hz': 20e6,
    'bw_hz': 20e6,
    'window': 'hamming',
    'overlap': 0.5,
    'K': 1,
    'size_planner': 'next_5smooth'
}

## Plan Capture Parameters

The `plan_capture_parameters()` function calculates optimal signal acquisition parameters based on the desired specifications. It takes the following inputs:

- rbw_hz: Desired Resolution Bandwidth in Hz (target frequency resolution)
- fs_hz: Sampling rate in Hz  
- bw_hz: Bandwidth in Hz
- window: Window function type (e.g. hamming)
- overlap: Overlap ratio between segments (0 to 1)
- K: Averaging factor
- size_planner: Method for determining FFT size (e.g. next_5smooth)

The function determines parameters like:
- Number of samples needed
- Actual achieved RBW
- Total capture time
- Optimal FFT size

This ensures efficient signal capture while meeting the specified frequency resolution requirements.


In [3]:
plan_capture_parameters(**plan)

{'sample_rate': 20000000.0,
 'samples': 27648,
 'rbw_eff': 983.7962962962964,
 'time': 0.0013824}

The acquisition plan parameters show:

- **Sample Rate**: 20 MHz (20,000,000 samples per second)
- **Number of Samples**: 27,648 samples per capture
- **Effective Resolution Bandwidth**: ~984 Hz (actual achieved RBW)
- **Capture Time**: 1.3824 milliseconds (total time to collect samples)

This plan achieves close to the desired 1 kHz RBW target while optimizing the FFT size for computational efficiency.


## Plan Welch Parameters

The `plan_welch_parameters()` function determines the optimal parameters for Welch's method of power spectral density estimation. It takes the same input parameters as `plan_capture_parameters()`:

- rbw_hz: Desired Resolution Bandwidth in Hz
- fs_hz: Sampling rate in Hz
- bw_hz: Bandwidth in Hz  
- window: Window function type
- overlap: Overlap ratio between segments
- K: Averaging factor
- size_planner: FFT size planning method

The function calculates the optimal parameters for spectral analysis including:
- NFFT: FFT size
- Segment size
- Number of segments
- Amount of overlap between segments
- Resulting frequency resolution

This ensures accurate spectral density estimation while meeting the specified resolution requirements.


In [4]:
plan_welch_parameters(**plan)

{'window': 'hamming',
 'nperseg': 27648,
 'noverlap': 13824,
 'nfft': 27648,
 'scaling': 'density',
 'average': 'mean'}

These Welch's method parameters define how the spectral analysis will be performed:

- **window**: Hamming window is used to reduce spectral leakage
- **nperseg**: 27,648 samples per segment - the length of each segment for FFT analysis 
- **noverlap**: 13,824 samples overlap between segments (50% overlap)
- **nfft**: FFT size of 27,648 points for frequency domain transformation
- **scaling**: 'density' returns power spectral density (power per frequency bin)
- **average**: 'mean' specifies arithmetic mean for averaging multiple segments

These parameters achieve:
- Frequency resolution of ~984 Hz
- Good tradeoff between spectral leakage and frequency resolution
- Efficient processing with 50% overlap between segments
